### PR TITLE
Move block creation constraints checker to BlockAPI

### PR DESCRIPTION
**Pull request by nzpr**
## Overview
<sup>_Currently node checks for constraints in BlockCreator, which means lots of computing happens before we check if node should create block at all. This PR moves checker closer to GRPC endpoint, inside BlockAPI._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Review by tgrospic — APPROVED** [review]
Nice work!

I updated Casper to have validator field as `Option[PublicKey]`. Checking if validator key is defined is then much more clear than checking for empty array. E.g. in [exploratory deploys](https://github.com/rchain/rchain/pull/2846/files#diff-2c412bf643d2b8a4b11e208d483e8710R618).

File: casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala [hunk 1]
```diff
@@ -124,17 +142,22 @@ class CreateBlockAPITest extends FlatSpec with Matchers with EitherValues {
...
-            _  <- n1.syncWith(n2)
+            b1                         <- n1.publishBlock(deploys(0))(nodes: _*)
+            b2                         <- n2.publishBlock(deploys(1))(nodes: _*)
+            _                          <- n1.syncWith(n2)
 
-            _        <- n1.casperEff.addDeploy(deploys(2))
+            _ <- n1.casperEff.addDeploy(deploys(2))
             b3Status <- createBlock(blockApiLock)(engineCell)
 
             _ = b3Status.left.value should include(
```
Comment (tgrospic) [inline casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala]: @nzpr this test is now failing. Is this expected?

Comment (nzpr) [reply casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala]: Thanks for fixing this! I guess I overlooked it.

Comment (tgrospic) [reply casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala]: I didn't look into test. What I meant is, with your changes should this test be adjusted or it's a newly created bug?

Comment (nzpr) [reply casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala]: test should be adjusted - `SynchronyConstraintChecker[Effect](0)` is 0 here, but it should be set to corresponding value https://github.com/rchain/rchain/blob/4997ae2fe829a7c7f199136542ea135e995be6f6/casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala#L141 in all tests. After I fixed this, now only the last test fails, looking into it.

Comment (nzpr) [reply casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala]: I will add ignore this test as now we are checking for constraints before check for deploys.

**Review by zsluedem — APPROVED** [review]
LGTM now

**Pull request merged**